File: rust/src/running_stat.rs
// Adapated from https://www.johndcook.com/blog/standard_deviation/
//               https://www.johndcook.com/blog/skewness_kurtosis/

#[derive(Clone, Copy)]
pub struct RunningStat {
    count: u32,
    old_mean: f64,
    new_mean: f64,
    old_sum: f64,
    new_sum: f64,
}

impl RunningStat {
    pub fn new() -> Self {
        Self {
            count: 0,
            old_mean: 0.0,
            new_mean: 0.0,
            old_sum: 0.0,
            new_sum: 0.0,
        }
    }

    pub fn clear(&mut self) {
        self.count = 0;
        self.old_mean = 0.0;
        self.new_mean = 0.0;
        self.old_sum = 0.0;
        self.new_sum = 0.0;
    }

    pub fn data_value_count(&self) -> u32 {
        self.count
    }

    pub fn mean(&self) -> f64 {
        match self.count > 0 {
            true => self.new_mean,
            false => 0.0,
        }
    }

    pub fn variance(&self) -> f64 {
        match self.count > 1 {
            true => self.new_mean / (self.count - 1) as f64,
            false => 0.0,
        }
    }

    pub fn standard_deviation(&self) -> f64 {
        self.variance().sqrt()
    }

    pub fn push(&mut self, value: f64) {
        self.count += 1;

        if self.count == 1 {
            self.old_mean = value;
            self.new_mean = value;
            self.old_sum = 0.0;
        } else {
            self.new_mean = self.old_mean + (value - self.old_mean) / self.count as f64;
            self.new_sum = self.old_sum + (value - self.old_mean) * (value - self.new_mean);

            // set up for next iteration
            self.old_mean = self.new_mean;
            self.old_sum = self.new_sum;
        }
    }
}

File: java/stat/RunningStat.java
import java.util.concurrent.locks.ReentrantLock;

class RunningStat {
    private int count = 0;
    private double oldMean = 0;
    private double newMean = 0;
    private double oldSum = 0;
    private double newSum = 0;

    private final ReentrantLock mutex = new ReentrantLock();

    public void RunningStat() {
        this.Clear();
    }

    public void Clear() {
        count = 0;
        oldMean = 0;
        newMean = 0;
        oldSum = 0;
        newSum = 0;
    }

    public int DataValueCount() {
        try {
            mutex.lock();
            return count;
        } finally {
            mutex.unlock();
        }
    }

    public double Mean() {
        try {
            mutex.lock();
            double mean = this.count > 0 ? this.newMean : 0;
            return mean;
        } finally {
            mutex.unlock();
        }
    }

    public double Variance() {
        try {
            mutex.lock();
            double variance = this.count > 1 ? this.newSum / (this.count - 1) : 0;
            return variance;
        } finally {
            mutex.unlock();
        }
    }

    public double StandardDeviation() {
        try {
            mutex.lock();
            double variance = this.Variance();
            return Math.sqrt(variance);
        } finally {
            mutex.unlock();
        }
    }

    public void Push(double value) {
        try {
            mutex.lock();
            this.count++;

            if (this.count == 1) {
                this.oldMean = value;
                this.newMean = value;
                oldSum = 0;
            } else {
                newMean = oldMean + (value - oldMean) / this.count;
                newSum = oldSum + (value - oldMean) * (value - newMean);
            }

            oldMean = newMean;
            oldSum = newSum;

        } finally {
            mutex.unlock();
        }

    }
}